source/v1/I2C.cpp

```diff
@@ -0,0 +1,229 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2015 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "mbed-drivers/v1/I2C.hpp"
+#include "minar/minar.h"
+#include "ualloc/ualloc.h"
+#include "core-util/CriticalSectionLock.h"
+#include "PeripheralPins.h"
+
+#if DEVICE_I2C && DEVICE_I2C_ASYNCH
+
+namespace mbed_drivers {
+using namespace mbed;
+namespace v1 {
+I2CTransaction::I2CTransaction(uint16_t address, uint32_t hz, bool irqsafe, I2C * issuer):
+    next(nullptr),
+    address(address),
+    root(nullptr),
+    current(nullptr),
+    hz(hz),
+    repeated(false),
+    irqsafe(irqsafe),
+    issuer(issuer)
+{}
+I2CTransaction::~I2CTransaction() {
+    current = root;
+    while (current) {
+        detail::I2CSegment * next = current->get_next();
+        issuer->free(current, irqsafe);
+        current = next;
+    }
+}
+
+detail::I2CSegment * I2CTransaction::new_segment()
+{
+    detail::I2CSegment * s = issuer->new_segment(irqsafe);
+    CORE_UTIL_ASSERT(s != nullptr);
+    if (!s) {
+        return nullptr;
+    }
+    s->set_next(nullptr);
+    mbed::util::CriticalSectionLock lock;
+    if (root == NULL) {
+        root = s;
+        current = s;
+    } else {
+        current->set_next(s);
+        current = s;
+    }
+    return s;
+}
+
+I2C::I2C(PinName sda, PinName scl) :
+    _hz(100000)
+{
+    // Select the appropriate I2C Resource Manager
+    uint32_t i2c_sda = pinmap_peripheral(sda, PinMap_I2C_SDA);
+    uint32_t i2c_scl = pinmap_peripheral(scl, PinMap_I2C_SCL);
+    ownerID = pinmap_merge(i2c_sda, i2c_scl);
+    int rc = detail::get_I2C_owner(ownerID)->init(sda, scl);
+    if (rc) {
+        ownerID = (size_t) -1;
```

Comment:
Is this an error state? Referenced in the line below : ``` if (ownerID == (size_t) -1) ```

It might make sense once we are doing a new i2c API version to define better return values?